// @sap/ux-specification/dist/src/sync/v4/export/controls/TableColumn.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
class ActionAfterExecutionConfiguration {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => application_1.getColumnsPathForLR(pageName, manifest, sectionId, columnKey) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfiguration.prototype, "navigateToInstance", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => application_1.getColumnsPathForLR(pageName, manifest, sectionId, columnKey) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfiguration.prototype, "enableAutoScroll", void 0);
exports.ActionAfterExecutionConfiguration = ActionAfterExecutionConfiguration;
class TableColumn {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    decorators_1.validity({
        since: '1.84.0'
    })
], TableColumn.prototype, "width", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    decorators_1.validity({
        since: '1.85.0'
    })
], TableColumn.prototype, "availability", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    decorators_1.validity({
        since: '1.84.0'
    })
], TableColumn.prototype, "afterExecution", void 0);
exports.TableColumn = TableColumn;
class TableCustomColumn {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    decorators_1.validity({
        since: '1.84.0'
    })
], TableCustomColumn.prototype, "header", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    decorators_1.validity({
        since: '1.84.0'
    })
], TableCustomColumn.prototype, "width", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    decorators_1.validity({
        since: '1.84.0'
    })
], TableCustomColumn.prototype, "template", void 0);
exports.TableCustomColumn = TableCustomColumn;
