// @sap/ux-specification/dist/src/specification/v4/controls/Table.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SelectionMode;
(function (SelectionMode) {
    SelectionMode["Multi"] = "Multi";
    SelectionMode["None"] = "None";
    SelectionMode["Single"] = "Single";
    SelectionMode["Auto"] = "Auto";
})(SelectionMode = exports.SelectionMode || (exports.SelectionMode = {}));
var TableTypeV4;
(function (TableTypeV4) {
    TableTypeV4["ResponsiveTable"] = "ResponsiveTable";
    TableTypeV4["GridTable"] = "GridTable";
    TableTypeV4["AnalyticalTable"] = "AnalyticalTable";
    /* Not supported yet:
    TreeTable = "TreeTable"*/
})(TableTypeV4 = exports.TableTypeV4 || (exports.TableTypeV4 = {}));
var TableCreationModeType;
(function (TableCreationModeType) {
    TableCreationModeType["NewPage"] = "NewPage";
    TableCreationModeType["Inline"] = "Inline";
    TableCreationModeType["CreationRow"] = "CreationRow";
})(TableCreationModeType = exports.TableCreationModeType || (exports.TableCreationModeType = {}));
var InitialLoadType;
(function (InitialLoadType) {
    InitialLoadType["Disabled"] = "Disabled";
    InitialLoadType["Enabled"] = "Enabled";
    InitialLoadType["Auto"] = "Auto";
})(InitialLoadType = exports.InitialLoadType || (exports.InitialLoadType = {}));
