// @sap/ux-specification/dist/src/sync/common/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ExportArtifacts;
(function (ExportArtifacts) {
    ExportArtifacts["flex"] = "flex";
    ExportArtifacts["manifest"] = "manifest";
})(ExportArtifacts = exports.ExportArtifacts || (exports.ExportArtifacts = {}));
var DirName;
(function (DirName) {
    DirName["Schemas"] = ".schemas";
    DirName["Pages"] = "pages";
})(DirName = exports.DirName || (exports.DirName = {}));
var FileName;
(function (FileName) {
    FileName["App"] = "app.json";
})(FileName = exports.FileName || (exports.FileName = {}));
exports.FacetTitlePrefix = 'Facet ID: ';
var SectionType;
(function (SectionType) {
    SectionType["Section"] = "Section";
    SectionType["HeaderSection"] = "HeaderSection";
})(SectionType = exports.SectionType || (exports.SectionType = {}));
var ArtifactType;
(function (ArtifactType) {
    ArtifactType["Manifest"] = "Manifest";
    ArtifactType["FlexChange"] = "FlexChange";
    ArtifactType["Annotation"] = "Annotation";
})(ArtifactType = exports.ArtifactType || (exports.ArtifactType = {}));
var ControlType;
(function (ControlType) {
    ControlType["Table"] = "sap.m.Table";
    ControlType["TableColumn"] = "sap.m.Column";
    ControlType["SmartTable"] = "sap.ui.comp.smarttable.SmartTable";
    ControlType["SmartFilterBar"] = "sap.ui.comp.smartfilterbar.SmartFilterBar";
    ControlType["SmartChart"] = "sap.ui.comp.smartchart.SmartChart";
    ControlType["SmartGroup"] = "sap.ui.comp.smartform.SmartGroup";
    ControlType["Button"] = "sap.m.Button";
    ControlType["ToolbarButton"] = "sap.m.OverflowToolbarButton";
    ControlType["Avatar"] = "sap.f.Avatar";
    ControlType["ObjectPageDynamicHeaderTitle"] = "sap.uxap.ObjectPageDynamicHeaderTitle";
    ControlType["ObjectPageGridProperties"] = "sap.ui.layout.GridData";
    ControlType["ObjectPageHeader"] = "sap.uxap.ObjectPageHeader";
    ControlType["ObjectPageLayout"] = "sap.uxap.ObjectPageLayout";
    ControlType["HeaderAction"] = "sap.uxap.ObjectPageHeaderActionButton";
    ControlType["DynamicPage"] = "sap.f.DynamicPage";
    ControlType["Form"] = "sap.ui.layout.form";
    ControlType["Chart"] = "sap.suite.ui.microchart";
    ControlType["Section"] = "sap.uxap.ObjectPageSection";
    ControlType["SubSection"] = "sap.uxap.ObjectPageSubSection";
})(ControlType = exports.ControlType || (exports.ControlType = {}));
var FacetBase;
(function (FacetBase) {
    FacetBase["LineItem"] = "LineItem";
    FacetBase["CollectionFacet"] = "CollectionFacet";
    FacetBase["Chart"] = "Chart";
    FacetBase["Form"] = "Form";
    FacetBase["Identification"] = "Identification";
    FacetBase["DataPoint"] = "DataPoint";
    FacetBase["Address"] = "Address";
    FacetBase["Contact"] = "Contact";
    FacetBase["Unknown"] = "";
})(FacetBase = exports.FacetBase || (exports.FacetBase = {}));
exports.BindingPropertyRegexAsString = "^{[A-Za-z0-9{}&$!@#? _|,<>'()[\\]\\/:=.]+}$";
var ViewTemplateType;
(function (ViewTemplateType) {
    ViewTemplateType["ResponsiveTableColumnsExtension"] = "ResponsiveTableColumnsExtension";
    ViewTemplateType["AnalyticalTableColumnsExtension"] = "AnalyticalTableColumnsExtension";
    ViewTemplateType["TreeTableColumnsExtension"] = "TreeTableColumnsExtension";
    ViewTemplateType["GridTableColumnsExtension"] = "GridTableColumnsExtension";
    ViewTemplateType["ResponsiveTableCellsExtension"] = "ResponsiveTableCellsExtension";
})(ViewTemplateType = exports.ViewTemplateType || (exports.ViewTemplateType = {}));
exports.METADATAPATH = 'webapp/localService/metadata.xml';
exports.MANIFESTPATH = 'webapp/manifest.json';
exports.VOCWITHSLASH = '/@com.sap.vocabularies';
exports.VOCWITHCOLONS = '::@com.sap.vocabularies';
exports.UIVOCABULARY = 'com.sap.vocabularies.UI.v1';
exports.QUICKVARPATH = '/quickVariantSelection';
exports.QUICKVARPATHX = '/quickVariantSelectionX';
exports.DATESETTINGSPATH = '/filterSettings/dateSettings';
exports.TOOLBAR = 'ToolBar<LineItems>';
exports.FRAGMENTNAMEPART = '.fragment.';
