// @sap/ux-specification/dist/src/sync/v4/export/pages/AnalyticalListPage.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const v4_1 = require("../../../../specification/v4");
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
// As there are no flex changes, this is not used. Need new logic when we have flex changes in V4
let AnalyticalListPage = class AnalyticalListPage {
};
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (entitySet) => application_1.getSettingsPathForPage(entitySet)
        }
    }),
    decorators_1.validity({
        since: '1.84.0'
    })
], AnalyticalListPage.prototype, "variantManagement", void 0);
__decorate([
    decorators_1.childId('fe::FilterBar')
], AnalyticalListPage.prototype, "filterBar", void 0);
__decorate([
    decorators_1.childId('fe::table')
], AnalyticalListPage.prototype, "table", void 0);
AnalyticalListPage = __decorate([
    decorators_1.pageLayoutInformation({
        id: v4_1.FE_TEMPLATE_V4_ALP,
        pageId: v4_1.FE_TEMPLATE_V4_ALP
    })
], AnalyticalListPage);
exports.AnalyticalListPage = AnalyticalListPage;
