// @sap/ux-specification/dist/src/sync/v2/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MULTISELECTTABLE = 'ResponsiveTableWithMultiSelect<LineItems>';
exports.INLINEDELETETABLE = 'ResponsiveTableWithInlineDelete<LineItems>';
exports.TREETABLE = 'TreeTable<LineItems>';
exports.ANALYTICALTABLE = 'AnalyticalTable<LineItems>';
exports.GRIDTABLE = 'GridTable<LineItems>';
var SectionTypeV2;
(function (SectionTypeV2) {
    SectionTypeV2["Section"] = "ObjectPageSectionV2";
    SectionTypeV2["SubSection"] = "ObjectPageSubSectionV2";
    SectionTypeV2["SectionForm"] = "ObjectPageSectionFormV2";
    SectionTypeV2["SectionTable"] = "ObjectPageSectionTableV2";
    SectionTypeV2["SectionChart"] = "ObjectPageSectionChartV2";
    SectionTypeV2["SubSectionForm"] = "ObjectPageSubSectionFormV2";
    SectionTypeV2["SubSectionTable"] = "ObjectPageSubSectionTableV2";
    SectionTypeV2["SubSectionChart"] = "ObjectPageSubSectionChartV2";
})(SectionTypeV2 = exports.SectionTypeV2 || (exports.SectionTypeV2 = {}));
